### PR TITLE
Add support for int types in gtScalar op.

### PR DESCRIPTION
**Pull request by pashu123**
Support for integer types in gtScalar op has been added.
The code share same logic with gtTensor op and can be merged
which is added as a TODO.

### CODE REVIEW

File: e2e_testing/torchscript/elementwise.py [hunk 1]
```diff
@@ -325,23 +325,40 @@ def ElementwiseMaximumModule_basic(module, tu: TestUtils):
 
 # ==============================================================================
 
-
-class ElementwiseGtScalarModule(torch.nn.Module):
+class ElementwiseGtScalarFloatModule(torch.nn.Module):
```
Comment (ramiro050) [inline e2e_testing/torchscript/elementwise.py:328]: nit: it seems like the convention below is to do `FloatScalar` rather than `ScalarFloat`

**Review by ramiro050 — APPROVED** [review]
LGTM! I just have a small style comment

File: e2e_testing/torchscript/elementwise.py [hunk 2]
```diff
@@ -86,23 +86,27 @@ def ElementwiseTernaryModule_basic(module, tu: TestUtils):
 
 
 class ElementwiseWhereSelfModule(torch.nn.Module):
-    def __init__(self):
-        super().__init__()
 
-    @export
-    @annotate_args([
-        None,
-        ([-1, -1, -1], torch.float32, True),
```
Comment (ramiro050) [inline e2e_testing/torchscript/elementwise.py:95]: why was this changed?

File: e2e_testing/torchscript/elementwise.py [hunk 3]
```diff
@@ -86,23 +86,27 @@ def ElementwiseTernaryModule_basic(module, tu: TestUtils):
...
-        ([-1, -1], torch.float32, True),
-        ([-1], torch.float32, True),
-    ])
-    def forward(self, a, b, c):
-        return torch.where(a > 0.5, b, c)
+  def __init__(self):
```
Comment (ramiro050) [inline e2e_testing/torchscript/elementwise.py:90]: nit: this file uses 4 spaces for indentation

**Review by ramiro050 — CHANGES_REQUESTED** [review]

File: e2e_testing/torchscript/elementwise.py (hunk 2 above)
Comment (pashu123) [reply e2e_testing/torchscript/elementwise.py:95]: `a>0.5` where a is of type float32 whereas 0.5 is to type float64. This is a type mismatch and aten.gt.Scalar would throw an error since type promotion is not handled there.

Comment (pashu123) [reply e2e_testing/torchscript/elementwise.py:95]: Previously, it wasn't promoting a to match type of (0.5) which is wrong, it should have promoted a.

Comment (ramiro050) [reply e2e_testing/torchscript/elementwise.py:95]: Type promotion is a very confusing and messy thing 😂 when it comes to scalars that are part of a tensor operation, if the scalar is a float, it gets converted to a `float32` (the default float type) and then the promotion takes place between the tensor and the scalar.

https://github.com/pytorch/pytorch/blob/deb6989880d8a62bb45ce6c6b058bb5d2e28cf91/aten/src/ATen/native/TypeProperties.cpp#L98-L105

If the scalar is part of an op that does not involve tensors, such as `NumToTensor` then the scalar does get converted to `float64` if it was a float scalar.

Therefore, for the case where the input tensor is a float tensor, all you have to do is convert the scalar to match the type of the float tensor.

Comment (ramiro050) [reply e2e_testing/torchscript/elementwise.py:95]: In general, it is not a good practice to remove a unit test once the unit test has been added. New changes should not break features that were already supported. I recommend undoing the changes you made for the float tensor types and try to get the lowering to work correctly with them

**Review by ramiro050 — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchToLinalg/TorchToLinalg.cpp [hunk 4]
```diff
@@ -1720,14 +1720,25 @@ static Value createLinalgPayloadCalculationForElementwiseOp(
...
+
+    // TODO: `gtTensor` and `gtScalar` share similar code and can be called from
+    // one static function.
+    Value otherPromoted =
+        convertScalarToDtype(b, loc, operands[1], payloadArgs[0].getType());
+
+    if (dtype.isa<mlir::FloatType>())
+      return b.create<arith::CmpFOp>(loc, arith::CmpFPredicate::UGT,
+                                     payloadArgs[0], otherPromoted);
+    if (IntegerType intType = dtype.dyn_cast<mlir::IntegerType>()) {
```
[the remark this answers is not in the document]
Comment (pashu123) [reply lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:1733]: Same is the case when tensor is of type int32 and scalar argument is of type int64? You are losing information in this case too.

File: e2e_testing/torchscript/elementwise.py [hunk 5]
```diff
@@ -339,10 +339,29 @@ def forward(self, x):
...
+        None,
+        ([-1, -1], torch.int64, True),
+    ])
+    def forward(self, x):
+        return torch.gt(x, 100)
+
+
+@register_test_case(module_factory=lambda: ElementwiseGtIntScalarModule())
+def ElementwiseGtIntScalarModule_basic(module, tu: TestUtils):
+    module.forward(torch.randint(3, (3,4)))
```
[the remark this answers is not in the document]
Comment (pashu123) [reply e2e_testing/torchscript/elementwise.py:362]: Cool. I will add that.

File: lib/Conversion/TorchToLinalg/TorchToLinalg.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (pashu123) [reply lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:1733]: @ramiro050 I have added more test cases. For the above case, I can add a TODO there, and that can be taken later.

File: e2e_testing/torchscript/elementwise.py [hunk 6]
```diff
@@ -339,10 +339,65 @@ def forward(self, x):
...
+    def __init__(self):
+        super().__init__()
+
+    @export
+    @annotate_args([
+        None,
+        ([-1, -1], torch.int64, True),
+    ])
+    def forward(self, x):
+        return torch.gt(x, 7.0)
```
Comment (ramiro050) [inline e2e_testing/torchscript/elementwise.py:375]: Here the important test is making sure that we are not rounding the float value to an integer. So you want to make sure that `torch.gt(x, 7.7)` and `torch.gt(x, 7.3)` work correctly. (they currently will not work correctly because the lowering is turning the float into an int).

It's okay to have features not supported, as long as they result in an unimplemented error. Since the functionality you want right now is `torch.gt(int, int)`, I would just make sure in the `int` branch of the lowering that both the scalar and the tensor are integers. Then leave this case as unimplemented and for a future patch to solve.

**Review by ramiro050 — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchToLinalg/TorchToLinalg.cpp [hunk 7]
```diff
@@ -1720,14 +1720,25 @@ static Value createLinalgPayloadCalculationForElementwiseOp(
   }
 
   if (auto gtScalar = dyn_cast<AtenGtScalarOp>(op)) {
-    Type dtype = gtScalar.self().getType().cast<ValueTensorType>().getDtype();
-    if (!dtype.isa<mlir::FloatType>()) {
-      gtScalar.emitError("unimplemented: non-floating point operand dtype");
-      return nullptr;
+    Type dtype = gtScalar.self().getType().cast<BaseTensorType>().getDtype();
+
+    // TODO: `gtTensor` and `gtScalar` share similar code and can be called from
```
Comment (silvasean) [inline lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:1725]: One side comment thought: Would it make sense to add a rule to ReduceOpVariants to convert `aten.foo.Scalar(x, y)` to `aten.foo.Tensor(x, NumToTensor(y))`? I think we can detect this case and add a trait "ScalarVariantWithCorrespondingTensorVariant" to the ops that have this property. It would look similar to how we handle emit_with_mutating_variants in https://github.com/llvm/torch-mlir/blob/528354de84c5892d539fce17c74b2c928a645991/python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/torch_ods_gen.py#L430

We would probably rename emit_with_mutating_variants -> emit_elementwise_op and have it detect which mutating/scalar/tensor variants exist, emit them, and add all the necessary traits.


Will let Ramiro drive this review, which seems to be going in a good direction anyway. Please submit after he gives approval.

**Review by silvasean — COMMENTED** [review]
Will let Ramiro drive this review, which seems to be going in a good direction anyway. Please submit after he gives approval. I left a comment about generalizing this though.

File: lib/Conversion/TorchToLinalg/TorchToLinalg.cpp (hunk 7 above)
Comment (ramiro050) [reply lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:1725]: I think adding something like this would be really nice. We just have to be careful about the type conversions that happened when a scalar is turned into a tensor. In particular, `NumToTensor` would have the wrong behavior because, for example, it converts scalar floats to `float64`, which would mess with the type promotion when performing the tensor computation. What we want is something that behaves like the `wrapped_number` in PyTorch, which follows a different set of type conversion rules.

File: lib/Conversion/TorchToLinalg/TorchToLinalg.cpp [hunk 8]
```diff
@@ -1720,14 +1720,32 @@ static Value createLinalgPayloadCalculationForElementwiseOp(
...
+        return b.create<arith::CmpIOp>(loc, arith::CmpIPredicate::ugt,
+                                       payloadArgs[0], otherPromoted);
+      if (intType.isSigned())
+        return b.create<arith::CmpIOp>(loc, arith::CmpIPredicate::sgt,
+                                       payloadArgs[0], otherPromoted);
     }
-    Value otherPromoted = convertScalarToDtype(b, loc, operands[1], dtype);
-    return b.create<arith::CmpFOp>(loc, arith::CmpFPredicate::UGT,
-                                   payloadArgs[0], otherPromoted);
+    gtScalar.emitError("unimplemented: dtype isn't supported.");
```
[the remark this answers is not in the document]
Comment (pashu123) [reply lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:1748]: done

File: lib/Conversion/TorchToLinalg/TorchToLinalg.cpp (hunk 7 above)
Comment (silvasean) [reply lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:1725]: +1, yeah, there are probably some subtle cases.
